feather/plugin-host/src/command_dispatcher.rs

```diff
@@ -0,0 +1,163 @@
+        // single cmd (to check that its regex does not use any unimplemented features)
+        /// by calling  NFA::<CmdPos<CommandId>>::from_command_regex(&cmd.regex,0).is_ok();
+        /// or (option b) add a method to lieutenants NFA that checks if the regex
+        /// contains any unimplemented features. See lieutenant/regex/regex_to_nfa, and
+        /// call that method instead. Anyways for these options we need someway of triggering
+        /// dfa updates. Could probably happen inside plugin load function, in main, or a host-call.
+        self.update_dfa();
+    }
+
+    // Can fail if we run out of u32 ids for the nodes in the nfa graph. This should be unlikely.
+    pub fn update_dfa(&mut self) -> anyhow::Result<()> {
+        let mut nfa = NFA::<CmdPos<CommandId>>::empty();
+
+        for (id, cmd) in self.commands.iter().enumerate() {
+            let cmd_nfa = NFA::<CmdPos<CommandId>>::from_command_regex(
+                &cmd.regex,
+                id.try_into()
+                    .expect("Cant register 2^32 different commands!"),
+            )?;
+            nfa = nfa.or(cmd_nfa)?;
+        }
+        // Reduses the number of states/nodes, so it cant fail.
+        self.dfa = nfa.into_early_termination_dfa();
+        Ok(())
+    }
+
+    /// calls the corresponding command (if it exists) and returns the resulting u32
+    /// Caller is player calling the command, None means terminal.
+    pub fn call(
+        &self,
+        game: &mut Game,
+        input: &str,
+        caller: Option<EntityId>,
+    ) -> anyhow::Result<i64> {
+        match self.dfa.early_termination_find(input) {
+            Ok(ids) => {
+                for id in ids {
+                    let description = &self.commands[id as usize];
+
+                    let plugin_manager =
+                        Rc::clone(&*game.resources.get::<Rc<RefCell<PluginManager>>>()?);
+                    let plugin_manager = plugin_manager.borrow();
+                    let plugin = plugin_manager.plugin(description.plugin_id);
+
+                    match plugin {
+                        Some(plugin) => {
+                            match plugin.call_command(
+                                game,
+                                input,
+                                description.callback.clone(),
+                                caller,
+                            ) {
+                                Ok(x) => {
+                                    // Command sucsessfully ran
+                                    return Ok(x);
+                                }
+                                Err(_) => {
+                                    // Was not able to call command, probably due to failed parsing.
+                                    // Unless there is some failure case i have not considerd, or someone
+                                    // is not using lieutenant, but their own shenanigans.
+                                    continue;
```

Comment:
Maybe we could have custom `bail!`s for those cases as well, instead of falling back to the default return.